router/src/completion.rs

```diff
@@ -0,0 +1,581 @@
+        text: resp.generated_text.clone(),
+        finish_reason: finish_reason,
+        logprobs: None,
+        index: 0,
+    };
+    let usage = Some(Usage {
+        completion_tokens: gen_tokens,
+        total_tokens: gen_tokens + prefill_len,
+        prompt_tokens: prefill_len,
+    });
+    let created_time = std::time::SystemTime::now().duration_since(std::time::UNIX_EPOCH).unwrap().as_secs() as u32;
+    let model = info.0.model_id;
+    let resp: CompletionsResponse = CompletionsResponse {
+        choices: vec![choices],
+        created: created_time,
+        id: String::from(format!("cmpl-{}", created_time)),
+        object: String::from("text_completion"),
+        model, 
+        usage,
+    };
+    Json(resp.into())
+}
+
+pub(crate) async fn generate_to_chatcompletions(
+    resp: Json<GenerateResponse>,
+    info: Extension<Info>
+) -> Json<ChatCompletionsResponse> {
+    // let details = resp.details.as_ref().ok_or("details missing"); //;
+    let details = resp.details.as_ref();
+    
+    let gen_tokens = match details {
+        Some(details) => details.generated_tokens,
+        None => 0,
+    };
+    let finish_reason = match details {
+        Some(details) => Some(details.finish_reason.clone()),
+        None => None,
+    };
+    let prefill_len = match details {
+        Some(details) => details.prefill.len() as u32,
+        None => 0,
+    };
+
+    let choices = ChatCompletionChoices {
+        message: ChatMessage { role: ChatRole::Assistant, content: resp.generated_text.clone() },
+        finish_reason: finish_reason,
+        index: 0,
+    };
+    let usage = Usage {
+        completion_tokens: gen_tokens,
+        total_tokens: gen_tokens + prefill_len,
+        prompt_tokens: prefill_len,
+    };
+    let created_time = std::time::SystemTime::now().duration_since(std::time::UNIX_EPOCH).unwrap().as_secs() as u32;
+    let model = info.0.model_id;
+    let resp = ChatCompletionsResponse {
+        choices: vec![choices],
+        created: created_time,
+        id: String::from(format!("chatcmpl-{}", created_time)),
+        object: String::from("chat.completion"),
+        model, 
+        usage,
+    };
+    Json(resp.into())
+}
+
+pub(crate) fn chat_start_message(created_time: u32) -> Box<dyn Send + erased_serde::Serialize> {
+    let choices: ChatCompletionDeltaStreamChoices = ChatCompletionDeltaStreamChoices { 
+        delta: ChatDeltaStreamMessage {content: None, role: Some(ChatRole::Assistant)},
+        finish_reason: None,
+        index: 0,
+    };
+    Box::new(ChatCompletionsStreamResponse {
+        choices: vec![choices],
+        created: created_time,
+        id: String::from(format!("chatcmpl-{}", created_time)),
+        object: String::from("chat.completion.chunk"),
+        model: String::from("TGIDummyChatCompletion"),
+    })
+}
+
+pub(crate) fn patch_streaming_event (st: StreamResponse, stream_type: StreamType, created_time: u32) -> Box<dyn Send + erased_serde::Serialize> {
```

Comment:
I don't understand the purpose of `patch_streaming_event`. Could you add some comment about what it does?

Reply:
If needed it reformats the StreamResponse with another protocol. It takes a StreamResponse 
and may return as output: 

- StreamResponse 
- ChatCompletionsStreamResponse 
- CompletionsResponse

Reply:
It's the only other viable way to switch waht the SSE returns. 
Any other approach would be to duplicate 

async fn generate_stream() per protocol, which would be highy redundant and more error prone in my view.

Reply:
I would argue that duplicating might be good for modularity if they are different. For example, if we change one protocol in the future, we would not affect other protocols.

Reply:
Makes also sense to me. 
Originally, I was looking for a way to include all in completion.rs 

That would make the code also more modular, changes would take some days